### ISSUE TITLE
Levels disappear 

### ISSUE THREAD

**Issue by artisteacher** [opentoonz/opentoonz#2581]
### Issue Summary

Occasionally levels are missing when a scene is opened, after starting OpenToonz. When this happens, the level is still associated with the scene and the column displays with red text. Looking in the browser, the actual file is no longer there. The "backup animation levels when saving" preference is enabled and sometimes the backup is missing too. In case it makes a difference, this is using an alternate project location (usually Documents) and using all of the default folders (drawings, extras & etc.).  Most likely this happens after a crash, but it is infrequent and so is hard to track.

### System Information
* **OpenToonz Version**: 1.3
* **Operating System**: macOS 10.13.6
* **CPU**: 
* **Memory**: 
* **Graphics Card**: 
* **Graphics Tablet**:

**Comment by RodneyBaker** [comment]
A likely problem with missing files and where the files are not located inside the stuff directory is that the source files have been moved.  Short of refraining from using files outside of the project directory structure it is not clear how best to proceed to minimize chance for loss of referenced files. 

One option might be to create a 'Secure Mode' in Preferences designed to intentionally restrict the user from performing risk without expressly acknowledging that risk.  This would be a long term developmental track with a primary goal to reduce variation.

With regard to Levels themselves disappearing there are several possibilities:

- Source was referenced in saved Scene but never saved itself
- Source was overwritten by another file an then subsequently updated out of scope (therefore no longer being the same name as the original sourced file
- Source was moved
- Source was corrupted
- Program, Project, Scene was moved relative to source file (source is still in same location but things that once referenced it have changed or moved
- Other things not itemized here

**Comment by artisteacher** [comment]
I agree that sometimes it must be user error due to things being moved or from saving the scene without saving levels. Sometimes though, there does seem to be something else going on. I know that I almost never move things from a project and I always save all (and have auto-save enabled). Perhaps it's the timing of saving coinciding with a crash. Does the scene file save before the levels with save all?

**Comment by RodneyBaker** [comment]
Levels can save from a number of different triggers.
The Save All is just the most efficient one observed thus far.

When considering loss of files I focus on risk.
We gained a great deal of freedom when Opentoonz could save outside of the stuff directory but it came with some added risk in that files outside of the stuff directory can have many things happen to them which will cause problems.

I don't see this as being in the realm of user error.
It is a referencing and processing problem.
It's not unlike all the dead internet links that occur after time.

What we could do is categorize the risks and attack them from that angle.
Speaking from my own workflow experience... I rarely start off a new session in Opentoonz by creating a new Project.  This is the first place where we enter risky territory because the default/fallback is to the Sandbox.  New users don't know the purpose of the sandbox nor do they understand how Projects work.

And that's before we even get to creating scenes and levels...

**Comment by RodneyBaker** [comment]
The File corruption risk has been demonstrated several times and it can still be demonstrated in some obscure instances.  More than just corruption though... Opentoonz sometimes does remove files... They are just gone.

The recent update to Backups and cache should hopefully address some of the inadvertent losses but they don't really drive to the root of the problem.

This is an area of interest that we should look at very closely.

**Comment by RodneyBaker** [comment]
If this issue relates to image types processed by ffmpeg it may be that #2876 and #2880 are a remedy for the missing levels.

**Comment by artisteacher** [comment]
Usually this happens with drawings created directly in OT, so the FFMPEG issues should not apply.

**Comment by manongjohn** [comment]
Are there particular file types this is happening with? Raster based files vs vector files?  

Similar to the ffmpeg issue, certain file types use an interim temp file to save the changes, delete the original and then rename the temp file to replace the original.  If it fails to create it for some reason, it will still delete the original but fail when it tries to rename the temp file that doesn't exist.  It would throw a rename error if you did a Save Level, but may not throw an error if you did a Save Scene but either way the original file is lost.

We should probably put in a logic to block the delete of the original file before the temp rename, if the temp file doesn't exist.  At least you shouldn't lose the original file.  You may lose the changes you made.

I'm a little concerned that a backup was not created or was lost.  Makes me wonder if there is some obscure bug with renaming files in OT's processing which could also explain the save issue.

**Comment by artisteacher** [comment]
> We should probably put in a logic to block the delete of the original file before the temp rename, if the temp file doesn't exist.

Great idea!

I know this has definitely happened with vector levels. I can't specifically remember if it has happened with raster, but it may have.

**Comment by RodneyBaker** [comment]
@artisteacher  If you can keep an eye on this report given the recent PRs that have been merged it will be appreciated.  Most of those PRs don't specifically relate to vector levels but the fixes there may be enough to close this report in anticipation of a new report that more specifically identifies,  describes and narrows down any remaining issues.

**Comment by artisteacher** [comment]
It literally happened again today - but I haven't updated the student computers to the latest nightly build in a while. I think this should stay open for now - at least until "Save All" triggers a failure notification too. 

> NOTE: At present time, failed level saves only notify the user if they did a "Save Level". "Save All" hides the failure notification. This should be addressed separately.

**Comment by RodneyBaker** [comment]
Thanks!
I note that the uncommenting of the code that saves newly generated scenes via Startup popup has yet to take place... @turtletooth anticipated the need and @shun-iwasawa confirmed the code is there (and is commented out) however.  I'm still a little concerned with the lack of initial scene save.  That shouldn't really effect the levels but anything that increases the chance of file loss certainly gets my attention.

I had an interruption of the very last part of a save earlier today where when I reopened the scene 1 out of 4 columns in the xsheet was missing (links in red).   It's not clear to me what caused other than I was testing PRs at time and that might account for the instability.  It hasn't reoccurred since.

**Comment by artisteacher** [comment]
This happened again last week when I was out sick. It sounds like the scene itself had been corrupted and wouldn't open. The student must have managed to get the backup open, but the levels were all missing. She did have backups enabled in preferences but there were no backup levels to be found. I'm hoping the update to 1.4 will prevent this from happening again, but I have noticed that it doesn't make backups for unedited levels if there are no backups #3091. To prevent unexpected file loss, backup levels should be created with Save All if they don't already exist, even if the level wasn't edited.

**Comment by phil123456** [comment]
opentoonz is just GARBAGE
I wasted a whole day of work

**Comment by RodneyBaker** [comment]
@phil123456 Assuming you are up to the challenge, if you need help please reach out and we'll be glad to assist.

**Comment by RodneyBaker** [comment]
Will move this to Opentoonz-docs as a reference for how references to levels/images can be broken resulting in the red textual reference in timeline/xsheet.
